### ISSUE TITLE
Are the lb_policy and lb_config forced to match?

### ISSUE THREAD

**Issue by DiazAlan** [envoyproxy/envoy#25583]
*Description*:
Currently in clusterInfoImpl when the lb_configs are loaded, it doesn't check if the lb_policy matches the config that's loaded. We have a separate member for the lb_policy and other members for each config type. Is there anything to stop the user from setting a round_robin config but also setting the lb_policy to least request? If there isn't, should anything be done about that?

*Relevant Links*:
[Current implementation of how the lb config members are initialized](https://github.com/envoyproxy/envoy/blob/0737bac16fa47ebe6a9c68d16a6c3244eb7ec051/source/common/upstream/upstream_impl.cc#L1011)

[Current location of where lb_policy is set](https://github.com/envoyproxy/envoy/blob/0737bac16fa47ebe6a9c68d16a6c3244eb7ec051/source/common/upstream/upstream_impl.cc#L1081)

**Comment by adisuissa** [comment]
cc'ing @wbpcode who did some work in this area recently.
It seems to me that it may be better to have a single owning field that states the correct lb_policy and uniquely determines the config that can be used. WDYT?

**Comment by wbpcode** [comment]
I agree a single field for load balancing would be great. And this is why we add the `load_balancing_policy`. This new field is recommended for normal load balancing (e.g. random, least request, etc.).

And we will migrate all current lb policy to this new field.
(these old fields still could be used in a long time for backward compatibility.

**Comment by adisuissa** [comment]
I think you are referring to the LB policy API, whereas OP is interested in whether the code in [upstream_impl] should be refactored to have a single data member that determines the load-balancing policy and its configuration.

**Comment by wbpcode** [comment]
> Is there anything to stop the user from setting a round_robin config but also setting the lb_policy to least request? If there isn't, should anything be done about that?

Sorry. I misunderstant it.
To the proto API, I don't think we can do anything. Any new validation or check is not allowd because of the backward compatibility.
And for the code in the `upstream_impl.h/cc`, IMO, merge the policy enum and the policy config make no big difference.

But If we can update the current implementation, to load policy config according to the value of policy enum would be helpful.
At least, we can ensure the policy configs match the policy enum values in the `upstream_impl.h/cc`. 🤔

**Comment by adisuissa** [comment]
> But If we can update the current implementation, to load policy config according to the value of policy enum would be helpful. At least, we can ensure the policy configs match the policy enum values in the `upstream_impl.h/cc`. 🤔

Yeah, I agree.
At the moment [lb_config](https://github.com/envoyproxy/envoy/blob/95a3faf69e8eef6edde867a3055ff4a5a9cd791a/api/envoy/config/cluster/v3/cluster.proto#L1054) may contain config type some LB-type, while [lb_policy](https://github.com/envoyproxy/envoy/blob/95a3faf69e8eef6edde867a3055ff4a5a9cd791a/api/envoy/config/cluster/v3/cluster.proto#L842) may contains a different LB-type value.

**Comment by DiazAlan** [comment]
Should I set the LBConfig in upstream_impl using the value of the lb_policy enum?

**Comment by wbpcode** [comment]
Sorry for the delay. I missed you question here. The answer is yes.

**Comment by wbpcode** [comment]
Changed my goal priorities. I will try to land the feature of `FilterConfig` first. 🤠
